### ISSUE TITLE
Replace lhash

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#10656]
The lhash algorithm is quite clever, in that it dynamically resizes both increasing and decreasing in size automatically.  It also amortises any size changes over a period of time by incrementally reallocating objects to buckets.  However, it isn't ideal for OpenSSL.

The problems are multi-fold:
1. We're stuck with the API for compatibility reasons;
2. It uses a cache unfriendly collision resolution mechanism (linked list of objects) and
3. OpenSSL almost never removes entries from of a hash table.

It would seem prudent to replace lhash which a new hsh table implementation that:
1. has a private but still type safe API;
2. behaves in a cache friendly manner and
3. is optimised for look up rather than deletion and insertion (specifically, deletion can be very expensive).

One possibility that meets these criteria is [hopscotch hashing](https://en.wikipedia.org/wiki/Hopscotch_hashing).  It isn't the only possibility.

An aside: there are lock free hash table algorithms that use atomic swap operations to guarantee consistency.  Hearsay evidence indicates that these aren't as efficient as locking equivalents in general.

Ideally, more than one hash table implementation would be included for different circumstances.  E.g. something that is essentially read only (mapping a name to an object or NID) could be implemented with no locking and more effort spent on removing collisions (q.v. [perfect hashing](https://en.wikipedia.org/wiki/Perfect_hash_function) which can be arranged in O(n) time, possibly at compile time).

Writing a new hash function or three is straightforward, going through the code base and appropriately retrofitting them everywhere will be time consuming.

**Comment by richsalz** [comment]
Do any of the public API's expose an LHASH?  If so, that would also be something to be addressed.

This should not happen for 3.0

**Comment by paulidale** [comment]
Yes :(  That's in addition to LHASH itself.

**Comment by paulidale** [comment]
A lot of the uses are internal which can be updated.  There is no way LHASH can be removed anytime soon.

**Comment by richsalz** [comment]
We should do some profile/timing tests for a "typical" webserver and see if this is worth doing. Defining the test benchmark will be hard.

**Comment by paulidale** [comment]
Yeah, demonstrating a clear benefit will be the challenge.  For a marginal improvement it wouldn't be worth the effort.  Fortunately, there are plenty of comparisons between various hashing schemes (e.g. [a comparison of major hashes](https://tessil.github.io/2016/08/29/benchmark-hopscotch-map.html), google finds many).  Hopscotch isn't universally preferred -- [this writer prefers Robin Hood](https://martin.ankerl.com/2016/09/15/very-fast-hashmap-in-c-part-1/).

There are non-locking possibilities too. For example: [concurrent hashes](https://dl.acm.org/citation.cfm?doid=3314574.3309206).  Perfect hashes could be a serious contender as well, especially if the cost can be paid at build time.  [gperf](https://www.gnu.org/software/gperf/) is one such and it generates hash functions that typically don't parse the entire key (although there is always a `strcmp(3)` for confirmation).

The current lhash isn't optimised for speed, it's optimised for automatic resizing.

**Comment by levitte** [comment]
I was tempted to build a hash API on top of SA (sparse array). Is this something you're considering, or are you thinking of a completely separate implementation?

**Comment by paulidale** [comment]
I was thinking separate.  Currently, lhash runs with a high load factor and sparse array is best for low density data.  Still, it is an interesting idea.

Sparse array access time is O(log n) for n = max index value.  A hash table is O(1).

**Comment by levitte** [comment]
> A hash table is O(1).

Ah, right, good point.

**Comment by richsalz** [comment]
If the hashtable is 1% of a typical SSL connection, there is no point in optimizing that.

Yes we can build a faster hash, but we need to see if it is worth it.

And since 3.0 already slipped once, I hope nobody will touch this during the current dev cycle.

**Comment by beldmit** [comment]
Is this bug still relevant?

**Comment by t8m** [comment]
IMO this is absolutely Post 3.0.0 stuff.

**Comment by levitte** [comment]
Re "worth the effort", I would factor in maintenance costs, i.e. what does it take to maintain "old klunky" compared to newer and fresher, and hopefully implemented with the lessons from the old in mind.

Also, I and I know that there are people who view this differently, I don't think that providing a general hash table implementation is OpenSSL's "core business", so if it were up to me, I'd make the new implementation internal.

**Comment by paulidale** [comment]
I concur with the idea of making this internal.  Our API guarantees make this difficult.  This is definitely post 3.0.

While lhash was nice at the time, we're not using it effectively.  It's designed to dynamically resize when adding **and removing** entries.  We almost always just add entries.  It's optimised for a different usage.

Add in that the state of the art has progressed quite a bit (cache friendly and lock free hash tables are both known now).  I'm confident we could do better.  Just not for 3.0 and not exported publicly.

**Comment by t8m** [comment]
We now have the new hashtable which can be used in places where it improves the performance over the old LHASH. I am closing this issue as individual replacements need to be considered and tracked separately.
